### ISSUE TITLE
US-023: Implement Advanced ML Models (LSTM, XGBoost, ARIMA)

### ISSUE THREAD

**Issue by cuauhtemocbe** [cuauhtemocbe/btc-predictor#25]
# US-023: Implement Advanced ML Models (LSTM, XGBoost, ARIMA)

## User Story

```
As a data scientist improving the BTC predictor
I want to implement multiple ML models (LSTM, XGBoost, ARIMA)
In order to compare which model generates the most accurate predictions and best returns
```

## Technical Context

**Current state:**
- Only one model implemented: `LinearRegressionModel`
- Abstract `BaseModel` class already exists in `workers/daily/models/base.py`
- Models table supports storing multiple serialized models
- Easy to add new models by inheriting from BaseModel

**Desired state:**
- Four models available: Linear Regression, LSTM, XGBoost, ARIMA
- Each model with reasonable default hyperparameters
- All models serialize/deserialize correctly (stored in `models` table)
- Each model tested independently

**Why these models:**
- **Linear Regression** (current): Simple baseline, fast training
- **LSTM** (Long Short-Term Memory): Captures temporal patterns, good for sequences
- **XGBoost**: Gradient boosting, handles non-linear relationships
- **ARIMA**: Time series classic, assumes stationarity

**Affected files:**
- `workers/daily/models/lstm_model.py` — new LSTM implementation
- `workers/daily/models/xgboost_model.py` — new XGBoost implementation
- `workers/daily/models/arima_model.py` — new ARIMA implementation
- `workers/daily/models/__init__.py` — export all models
- `pyproject.toml` — add dependencies: `tensorflow`, `xgboost`, `statsmodels`
- `workers/daily/tests/test_models.py` — tests for each model

**Hyperparameters (reasonable defaults):**

```python
# LSTM
{
    "window_days": 30,
    "lstm_units": 50,
    "dropout": 0.2,
    "epochs": 50,
    "batch_size": 32
}

# XGBoost
{
    "window_days": 30,
    "n_estimators": 100,
    "max_depth": 5,
    "learning_rate": 0.1
}

# ARIMA
{
    "order": (5, 1, 0),  # (p, d, q)
    "seasonal_order": (0, 0, 0, 0)
}
```

**Edge cases (ZOMBIES):**
- **Z (Zero)**: Model trained with 0 samples → raise ValueError
- **O (One)**: Model trained with 1 sample → insufficient, raise error
- **M (Many)**: Large dataset (365+ days) → ensure training completes in <5 min
- **B (Boundaries)**: Min/max values for hyperparameters
- **I (Interfaces)**: All models implement same BaseModel interface
- **E (Exceptions)**: Training fails (OOM, convergence) → log error, re-raise
- **S (Serialization)**: Each model serializes/deserializes correctly (pickle/keras)

## Acceptance Criteria

```gherkin
Feature: Advanced ML models for BTC prediction

  Background:
    Given the BaseModel abstract class is defined
    And it requires implementing: train(), predict(), serialize(), deserialize()

  Scenario: LSTMModel implements BaseModel interface
    When I create an LSTMModel instance
    Then it has methods: train(X, y), predict(X), serialize(), deserialize(artifact)
    And it inherits from BaseModel
    And it can be imported from workers.daily.models

  Scenario: Train LSTM model with default hyperparameters
    Given historical BTC prices for 60 days (30 train + 30 test)
    And LSTM hyperparameters:
      | window_days | 30   |
      | lstm_units  | 50   |
      | dropout     | 0.2  |
      | epochs      | 50   |
    When I train the LSTM model
    Then training completes without errors
    And the model is fitted (has trained weights)
    And I can call predict(X) with shape (1, 30)
    And it returns a float price prediction

  Scenario: LSTM model serializes and deserializes correctly
    Given a trained LSTM model
    When I call serialize()
    Then it returns a bytes artifact (pickled Keras model)
    And I can save it to the database
    When I call LSTMModel.deserialize(artifact)
    Then it returns a new LSTMModel instance
    And predictions from the deserialized model match the original

  Scenario: XGBoostModel implements BaseModel interface
    When I create an XGBoostModel instance
    Then it implements train(), predict(), serialize(), deserialize()
    And it uses xgboost.XGBRegressor internally

  Scenario: Train XGBoost model with default hyperparameters
    Given historical BTC prices for 60 days
    And XGBoost hyperparameters:
      | window_days   | 30  |
      | n_estimators  | 100 |
      | max_depth     | 5   |
      | learning_rate | 0.1 |
    When I train the XGBoost model
    Then training completes in < 10 seconds
    And I can make predictions
    And predictions are in reasonable range (not NaN, not negative)

  Scenario: ARIMAModel implements BaseModel interface
    When I create an ARIMAModel instance
    Then it implements the BaseModel interface
    And it uses statsmodels.tsa.arima.ARIMA internally

  Scenario: Train ARIMA model with default order
    Given a time series of 60 daily BTC close prices
    And ARIMA order = (5, 1, 0)
    When I train the ARIMA model
    Then it fits an ARIMA(5,1,0) model
    And I can predict 1 step ahead
    And the prediction is a float price

  Scenario: ARIMA handles non-stationary data with differencing
    Given BTC prices with a clear trend (non-stationary)
    And ARIMA order = (5, 1, 0) with d=1 (first difference)
    When I train the model
    Then it applies differencing internally
    And predictions are reasonable (within 10% of recent prices)

  Scenario Outline: All models produce valid predictions
    Given a trained <model_name> model
    And input features X with shape (1, 30)
    When I call model.predict(X)
    Then it returns a single float prediction
    And the prediction is > 0 (price cannot be negative)
    And the prediction is within 50% of the last known price (sanity check)

    Examples:
      | model_name       |
      | LinearRegression |
      | LSTM             |
      | XGBoost          |
      | ARIMA            |

  Scenario: Models are stored with their hyperparameters in database
    Given I train an LSTM model with lstm_units=50
    When I save it to the models table
    Then the params JSONB column contains:
      ```json
      {
        "model_name": "lstm_v1",
        "window_days": 30,
        "lstm_units": 50,
        "dropout": 0.2,
        "epochs": 50
      }
      ```

  Scenario: Handle insufficient training data gracefully
    Given I have only 10 days of historical data
    And a model requires 30 days (window_days=30)
    When I try to train the model
    Then it raises ValueError with message "Insufficient data: need 30, have 10"
    And no model is saved to the database

  Scenario: LSTM training completes in reasonable time
    Given 90 days of training data
    And LSTM epochs=50
    When I train the model
    Then training completes in < 5 minutes
    And the model converges (loss decreases)

  Scenario: XGBoost handles large datasets efficiently
    Given 365 days of training data
    When I train XGBoost with n_estimators=100
    Then training completes in < 30 seconds
    And memory usage is < 500MB
```

## Definition of Done

- ✅ All Gherkin scenarios have passing automated tests
- ✅ LSTMModel, XGBoostModel, ARIMAModel all inherit from BaseModel
- ✅ Dependencies added to pyproject.toml: `tensorflow`, `xgboost`, `statsmodels`
- ✅ Each model has unit tests (train, predict, serialize/deserialize)
- ✅ All models can be stored in the `models` table
- ✅ Test coverage >= 85% for new model classes
- ✅ Serialization/deserialization tested for each model
- ✅ Lint and type-check green
- ✅ Documentation: docstrings for each model class

## Effort: L (4-5 days)

**Breakdown:**
- LSTM implementation + tests: 1.5 days
- XGBoost implementation + tests: 1 day
- ARIMA implementation + tests: 1 day
- Serialization/deserialization for all models: 0.5 day
- Integration tests (all models together): 0.5 day
- Documentation: 0.5 day

**Comment by cuauhtemocbe** [comment]
✅ **US-023 Completado**

Implementados tres modelos ML avanzados para predicción multi-modelo de BTC:

**Modelos:**
- ✅ XGBoostModel - Gradient boosting (100 estimators)
- ✅ LSTMModel - Neural network (50 LSTM units)
- ✅ ARIMAModel - Time series (order 5,1,0)

**Tests:**
- 24 tests XGBoost (todos passing)
- 26 tests LSTM
- 23 tests ARIMA
- Integration tests para todos los modelos

**Deployment:**
- ✅ Código pusheado a main (commit 5489b91)
- ✅ Railway deployment en progreso
- ✅ Dependencies instaladas: tensorflow-cpu, xgboost, statsmodels

**Archivos:**
- `workers/daily/models/xgboost_model.py`
- `workers/daily/models/lstm_model.py`
- `workers/daily/models/arima_model.py`
- Tests comprehensivos + integration tests

Todos los modelos heredan de BaseModel, implementan train/predict/serialize/deserialize, y se pueden almacenar en PostgreSQL.

🤖 Implementado por Claude Code siguiendo Spec-Driven Development workflow completo.
